pkg/uprobetracer/usdt.go

```diff
@@ -151,11 +349,130 @@ func getUsdtInfo(filepath string, attachSymbol string) (*usdtAttachInfo, error)
 			}
 		}
 
-		provider := readStringFromBytes(desc, uint32(3*wordSize))
-		probe := readStringFromBytes(desc, uint32(3*wordSize+len(provider)+1))
-		if provider == providerName && probe == probeName {
-			return &usdtAttachInfo{location, elfSemaphore}, nil
+		// parse arguments
+		argumentsString := readStringFromBytes(desc, uint32(3*wordSize+len(provider)+1+len(probe)+1))
+		arguments := &[usdtArgsMaxCount]uint64{}
+		for i, arg := range strings.Split(argumentsString, " ") {
+			arguments[i] = encodeArgument(arg)
 		}
+
+		results.attachAddresses = append(results.attachAddresses, location)
+		results.semaphoreAddresses = append(results.semaphoreAddresses, elfSemaphore)
+		results.arguments = append(results.arguments, arguments)
+	}
+
+	if len(results.attachAddresses) == 0 {
+		return nil, errors.New("no matching USDT metadata")
 	}
-	return nil, errors.New("no matching USDT metadata")
+
+	return results, nil
+}
+
+func getUsdtArgsInfoMap() (*ebpf.Map, error) {
+	if usdtArgsInfoMap != nil {
+		return usdtArgsInfoMap, nil
+	}
+
+	usdtArgsInfoMapSpec := ebpf.MapSpec{
+		Name:       UsdtArgsInfoMapName,
+		Type:       ebpf.Hash,
+		KeySize:    8,
+		ValueSize:  8 * usdtArgsMaxCount,
+		MaxEntries: 10000,
+	}
+	var err error
+	usdtArgsInfoMap, err = ebpf.NewMap(&usdtArgsInfoMapSpec)
+	if err != nil {
+		return nil, fmt.Errorf("creating USDT arguments info map: %w", err)
+	}
+	return usdtArgsInfoMap, nil
+}
+
+func GetUsdtArgsBufferMap() (*ebpf.Map, error) {
+	if usdtArgsBufferMap != nil {
+		return usdtArgsBufferMap, nil
+	}
+
+	usdtArgsBufferMapSpec := ebpf.MapSpec{
+		Name:       UsdtArgsBufferMapName,
+		Type:       ebpf.PerCPUArray,
+		KeySize:    4,
+		ValueSize:  8,
+		MaxEntries: 1,
+	}
+	var err error
+	usdtArgsBufferMap, err = ebpf.NewMap(&usdtArgsBufferMapSpec)
+	if err != nil {
+		return nil, fmt.Errorf("creating USDT arguments buffer map: %w", err)
+	}
+	return usdtArgsBufferMap, nil
```

Comment:
Today I learned `sync.Once()`, but it accepts a lambda so it's hard to get the err message if the map fails to be created?

Reply:
You could declare the `err` variable outside of the lambda to capture it.

Example of code doing that:
https://github.com/inspektor-gadget/inspektor-gadget/blob/36f4576187f1349fda491ccbddd24446e369311c/pkg/utils/host/namespaces.go#L33-L35

Are you suggesting to use a single maps for the gadget, or for all gadgets?

Reply:
> Are you suggesting to use a single maps for the gadget, or for all gadgets?

I'm using one map for all the gadgets. The map should be loaded with the ebpf extension, not the gadget. If we use individual maps for each gadget, we will need to load the extension for N times, and it will be much slower.